### PR TITLE
Add voter state workflow

### PR DESCRIPTION
**Pull request by alavieille**
[OO-FEATURE]  Add voter to check workflow state of statusable document

https://github.com/open-orchestra/open-orchestra-workflow-function-bundle/pull/97/files
https://github.com/open-orchestra/open-orchestra-cms-bundle/pull/1571

### CODE REVIEW

File: WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowStateVoterTest.php [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+        $this->user = Phake::mock('OpenOrchestra\UserBundle\Document\User');
+        Phake::when($this->user)->getId()->thenReturn('fakeUserId');
+        Phake::when($this->user)->getUsername()->thenReturn($this->username);
+
+        $this->token = Phake::mock('Symfony\Component\Security\Core\Authentication\Token\TokenInterface');
+        Phake::when($this->token)->getUser()->thenReturn($this->user);
+
+        $this->voter = new WorkflowStateVoter($this->workflowRightRepository, $this->contentTypeRepository);
+
```
Comment (KevinArtus) [inline WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowStateVoterTest.php]: Remove this lline

File: WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowStateVoterTest.php [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+     * @param array  $attributes
+     * @param array  $workflowRoles
+     * @param bool   $isOwner
+     * @param bool   $superAdmin
+     *
+     * @dataProvider provideObject
+     */
+    public function testVote($accessResponse, $object, $referenceId, array $attributes, array $workflowRoles, $isOwner = false, $superAdmin = false)
+    {
+
```
Comment (KevinArtus) [inline WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowStateVoterTest.php]: remove this line

File: WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowStateVoterTest.php [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        $status1 = Phake::mock('OpenOrchestra\ModelInterface\Model\StatusInterface');
+        $arrayCollectionDraftToPending = Phake::mock('Doctrine\Common\Collections\ArrayCollection');
+        $arrayCollectionPendingToTranslate = Phake::mock('Doctrine\Common\Collections\ArrayCollection');
+        Phake::when($status1)->getFromRoles()->thenReturn($arrayCollectionDraftToPending);
+        Phake::when($status1)->getToRoles()->thenReturn($arrayCollectionPendingToTranslate);
+        Phake::when($arrayCollectionDraftToPending)->toArray()->thenReturn(array($roleDraftToPending));
+        Phake::when($arrayCollectionPendingToTranslate)->toArray()->thenReturn(array($rolePendingToTranslate));
+
+
```
Comment (KevinArtus) [inline WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowStateVoterTest.php]: remove this line

**Comment by alavieille** [comment]
@KevinArtus Pr updated

File: WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowStateVoter.php [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        foreach ($attributes as $role) {
+            if (!$this->supportsAttribute($role)) {
+                return self::ACCESS_ABSTAIN;
+            }
+        }
+        if ($isOrchestraUser) {
+            $isOwner = $object instanceof BlameableInterface && ($object->getCreatedBy() == $user->getUsername());
+            $workflowRight = $this->workflowRightRepository->findOneByUserId($user->getId());
+            if (null === $workflowRight) {
+                return VoterInterface::ACCESS_DENIED;
```
Comment (ngilain) [inline WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowStateVoter.php:90]: I think if the user is the owner and the document is in ititial state, the voter shouldn't return access denied

Comment (alavieille) [inline WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowStateVoter.php:90]: @itkg-ngilain  This part is out of the scope, i create a ticket for speak of this problem

**Comment by veolys-nanne** [comment]
:+1:

**Pull request merged**